### ISSUE TITLE
[FEAT] Build hardened malloc into glibc

### ISSUE THREAD

**Issue by monsieuremre** [secureblue/secureblue#293]
I have examined the copr repo and the project itself. There is definetely room for improvement.

Suggestion 1: Explore other ways of enabling other than preloading
I see you opt in to enable hardened malloc for everything. In this case, it is not the best idea to preload the library when calling binaries. First of all, preloading is a patch. Second of all, it may not, and in some cases will not work, for child processes that are spawned by an existing process. A program can easily choose to scrub the envrionment before spawning or forking. In fact, for security purposes, it is a good practice to do this whenever possible. Also, if it is very trivial to disable a preloaded library for a process, if it has certain privileges. So your way of enabling it is absolutely miles ahead of not using a hardened allocator at all, but it can be improved. Compiling the gnu c library with a custom memory allocator from another repository can be done with some config changes.

But here the question is, how to replace the old package. We want all packages that depend on glibc to automatically check and pull our new library. I don't know how to do this on fedora. With apt, it is possible and easy tho. But even not this, we can still keep the old package but symlink the library directory to our own library or something. So this can be worked out.

I can try helping with this, if desired.

Suggestion 2: Make the copr repo standalone
The config file to enable hardened malloc is not part of the rpm package. It should be.

Suggestion 3: Where are exceptions?
Seriously. It is documented that crypt-setup works about 10 times slower with hardened malloc when decrypting a disk. How was this observed? Also, certain applications are known break with custom mallocs. There are workarounds. At the very least these applications should be exempted or workaround config files should be installed with the package.

**Comment by RoyalOughtness** [comment]
> Compiling the gnu c library with a custom memory allocator from another repository can be done with some config changes.
> But here the question is, how to replace the old package. We want all packages that depend on glibc to automatically check and pull our new library. I don't know how to do this on fedora. With apt, it is possible and easy tho. But even not this, we can still keep the old package but symlink the library directory to our own library or something. So this can be worked out.

True, this is the recommendation from @thestinger too. It needs further investigation.


> The config file to enable hardened malloc is not part of the rpm package. It should be.

Why does this matter?

> Where are exceptions?

Users can add exceptions based on their own preferneces. It is trivial to do so on a per application basis for flatpaks.

> At the very least these applications should be exempted or workaround config files should be installed with the package.

Maintaining a set of workarounds for all applications that exist is not in scope. It is also not necessary, since hardened_malloc works well with most applications and can be disabled as needed.

**Comment by monsieuremre** [comment]
> It needs further investigation.

I have already compiled glibc with my own primitive malloc implementation (neither more secure, nor faster, just wrote it when learning about memory allocation). It should not be that different to do it with hardened-malloc from grapheneOS. It is a clean project easy to compile. An integrated c library with hardened malloc as a concept already exists as a concept for bionic (and musl I think too). Glibc should work too. I might experiment with this.

> Why does this matter?

It does not, if you plan to bundle everything as some 'all in or all out' kind of project. But if you want the hardened-malloc package to be installable for other fedora users not on secureblue, for example those that are just on the normal fedora workstation, it might make more sense that the package is standalone and not entangled with other components from secureblue.

> Users can add exceptions based on their own preferneces. It is trivial to do so on a per application basis for flatpaks.

Yeah, maybe the right approach. It might be exhaustive to maintain an exception list. In fact, @cig0 had problems with firefox in https://github.com/secureblue/secureblue/issues/246 and I basically explained the reason of the problem and the solution (without an exemption from hardened malloc). The reason being 'jemalloc'. I see you deleted that comment for some reason. Am I missing something here? Everything I said there was just from the mozilla documentation directly and testable literally in seconds. I don't know which part of that came out as 'baseless misinformation' but you are free to debunk and correct me if I say something wrong instead of deleting the comment. I would like to learn the truth too if im wrong for that matter.

**Comment by thestinger** [comment]
Fewer applications will need exceptions with it built into libc since it will eliminate the need for exceptions to work around the bugs / limitations in glibc's malloc replacement support which is what's causing some of the issues. GrapheneOS supports exceptions with it built into libc by keeping the standard allocator there and supporting choosing between them via function pointer tables that are made read-only in early libc init. We simply reused the standard Android infrastructure for it. It does add a very tiny bit of overhead but in Bionic it was already there for other uses.

**Comment by RoyalOughtness** [comment]
@monsieuremre 

> I see you deleted that comment for some reason. Am I missing something here? Everything I said there was just from the mozilla documentation directly and testable literally in seconds.

That is not accurate. You claimed with zero supporting evidence that Firefox was now preferable security-wise to chromium-based browsers and that its sandboxing mechanisms had surpassed chromium's. Your efforts to downplay what you wrote are unconvincing. If you're going to make claims, substantiate them or don't make them at all.

**Comment by thestinger** [comment]
Firefox still doesn't have completed site isolation and the sandbox is still much weaker.

**Comment by RoyalOughtness** [comment]
@monsieuremre  

> thats not only a security comparison.

Then it doesn't belong here.

> let me substantiate it for you.

Substantiation means providing references. I will continue to delete your comments that spread misinformation as they risk misleading users. This is now the second time I'm requesting that you accept this constructive feedback. If you continue to reject it, it will lead to a temp or permanent ban per the code of conduct. Consider this a formal [warning](https://github.com/secureblue/secureblue?tab=coc-ov-file#2-warning).

> I have no data on that unfortunately.

You should be collecting that data before making these kinds of claims.

**Comment by monsieuremre** [comment]
> Firefox still doesn't have completed site isolation and the sandbox is still much weaker.

I am aware this argument could be made and it is valid, as I mentioned.

@qoijjj deleting my comments and still answering them is not my understanding of constructive criticism as it misrepresents what I actually said. I find it to be very unfortunate that you take this approach.

Anyway, I will respect your rules. You don't have to worry anymore because I do not plan to engage with this project. I am closing this issue for this reason. If you think this is a valid issue and you still want to pursue it, please open a follow up issue on your own so that I am no longer involved. Best of luck.

I will look into compiling a hardened malloc integrated glibc to work on arch. If I do that, I will comment it here so that you can port it to fedora if you think it suits your needs.

**Comment by pm4rcin** [comment]
> Firefox still doesn't have completed site isolation and the sandbox is still much weaker.

You're talking on desktop Firefox? I only know that Flatpak firefox is not as strong since no namespaces support for now. Would you provide some link or care to explain because I thought site isolation is mostly as Chrome except things like CFI etc.

**Comment by thestinger** [comment]
Yes, desktop Firefox has a **much** weaker sandbox implementation than Chromium across platforms.

It also still hasn't completed the implementation of site isolation, meaning sites and their data are still not properly protected from each other. They have a lot of the site isolation work done but it's incomplete. Look at their tracker. Without site isolation, the sandbox is mostly only protecting the OS rather than also browser data and sites from each other.

**Comment by secretmango** [comment]
@monsieuremre Fedora Firefox is compiled with `--enable-replace-malloc` so it works with hardened_malloc.

[Commit](https://src.fedoraproject.org/rpms/firefox/c/69510db0506067e3b711bdd4bef180812009286d)

[Updated Commit for x86_64 only](https://src.fedoraproject.org/rpms/firefox/c/fc0a23427b1fc5cc7f88ef11859106ea410c744b)

Without continuing the discussion on Firefoxes security, note
- This patch is only done in Fedora Firefox, does not work with Torbrowser or the mozilla binary
- an [rpm-ostree bug](https://github.com/coreos/rpm-ostree/issues/4554) prevents from installing previously removed packages
- [Fedora people want to switch to Flatpak Firefox](https://gitlab.com/fedora/ostree/sig/-/issues/3), where Firefox has no filesystem sandbox at all (due to a missing fork server, so it cannot use Zypak), and until now I didnt read any comments about security
- Librewolf can be installed but [their build system is pretty broken](https://codeberg.org/librewolf/issues/issues/1897) which they are just fixing, may be done

On Secureblue, currently users cannot really run Firefox. 

[Here is a discussion about the torbrowser and hardened_malloc](https://gitlab.torproject.org/tpo/applications/tor-browser-build/-/merge_requests/933) where the patch was rejected. They said themselves that the reasons (potential fingerprinting by performance) makes little sense, as all the hardware is different.

So this also affects when running Torbrowser through Bubblejail with userns-sandboxes enabled, as the RPM and their binary dont allow using a different malloc. You need to use the "Torbrowser Launcher" Flatpak, which, again, has no filesystem sandboxing due to a missing fork server in Firefox.

**Comment by RoyalOughtness** [comment]
> an https://github.com/coreos/rpm-ostree/issues/4554 prevents from installing previously removed packages

FYI there is a hacky workaround for this https://discord.com/channels/1202086019298500629/1202086019793702952/1265081906089693285

**Comment by secretmango** [comment]
> > an [coreos/rpm-ostree#4554](https://github.com/coreos/rpm-ostree/issues/4554) prevents from installing previously removed packages
> 
> FYI there is a hacky workaround for this https://discord.com/channels/1202086019298500629/1202086019793702952/1265081906089693285

I cant open that without a Discord account, could you copy it here? Thanks!

**Comment by monsieuremre** [comment]
@secretmango using the flatpak version of firefox would not be a good idea, nor using any browser with flatpak for that matter. In the case of chromium, it depends on SUID binaries and firefox depends on accessing various kernel features directly, to implement various security features. Flatpak makes this not possible.

I don't wanna get into detail because this discussion was already marked off-topic. But I do want to add one side note: the problem with flatpak and browsers is not because sandboxing the browser inherently weakens the security or anything. For example, if you use the snap version of chromium, you do not get the security hit that you get with flatpak. This is because snaps were designed to also confine high-privilege binaries and system services too and not just user apps. Just one example.

**Comment by secretmango** [comment]
@monsieuremre afaik only on the nouserns images does Chromium require suid. On the normal ones, as well as on ChromeOS, it uses user namespaces.

Snaps are just sandboxed with Apparmor, which is pretty different. This makes the sandboxing incompatible with Fedora for example.

The too strict seccomp filters on Flatpaks are the cause. Processes can use nested seccomp filters for more protection, but cant remove them. They are pretty loose (badness enumeration) but block userns creation.

/end

I would appreciate a discussion page here, as I understand that piling up issues is not nice.

**Comment by RoyalOughtness** [comment]
Discord is available for general discussion, this issue is about hardened malloc.

**Comment by RoyalOughtness** [comment]
I've been thinking about this a bit more.

Even with hardened_malloc built into libc, we will run into the issue of how to manage overrides for specific processes, and in such a way that processes can't impersonate other processes to switch themselves to a weaker malloc. On Android this piece is relatively trivial because apps are signed and can't impersonate each other.

Replicating this part and doing it right is the harder part of this feature. If anyone has ideas along these lines, I'm curious to hear them.

**Comment by nihil-admirari** [comment]
> But here the question is, how to replace the old package. We want all packages that depend on glibc to automatically check and pull our new library. I don't know how to do this on fedora.

[How to achieve “dnf swap”-equivalent functionality with rpm-ostree](https://fedoramagazine.org/how-to-achieve-dnf-swap-equivalent-functionality-with-rpm-ostree/). Haven't tried it myself.

**Comment by pm4rcin** [comment]
@thestinger a few days ago some people claimed that firefox has reached parity with Chromium in regard to sandboxing. Is there any merit for those claims? https://reddit.com/r/firefox/comments/1qkqfcx/firefox_sandbox_isolation_hits_level_9_the_gap/

**Comment by RoyalOughtness** [comment]
@pm4rcin This is not a discussion forum and that post is referring to process isolation on Windows.
